etc/systemd/system-generators/zfs-mount-generator.c

```diff
@@ -0,0 +1,1060 @@
+				fprintf(stderr, PROGNAME ": %s (%d): "
+				    "%s already exists. Skipping.\n",
+				    dataset, getpid(), mountfile);
+		}
+
+		// File exists: skip current dataset.
+		if (strcmp(p_canmount, "on") == 0)
+			sem_post(&noauto_files->noauto_names_sem);
+		return (0);
+	} else {
+		if (is_known) {
+			// See 4.
+			if (strcmp(p_canmount, "on") == 0)
+				sem_post(&noauto_files->noauto_names_sem);
+			return (0);
+		} else if (strcmp(p_canmount, "noauto") == 0) {
+			if (noauto_files->noauto_names_len ==
+			    noauto_files->noauto_names_max)
+				fprintf(stderr, PROGNAME ": %s (%d): "
+				    "noauto dataset limit (%zu) reached! "
+				    "Not tracking %s. Please report this to "
+				    "https://github.com/openzfs/zfs\n",
+				    dataset, getpid(),
+				    noauto_files->noauto_names_max, mountfile);
+			else {
+				strncpy(noauto_files->noauto_names[
+				    noauto_files->noauto_names_len],
+				    mountfile, NAME_MAX);
+				++noauto_files->noauto_names_len;
+			}
+		}
+	}
+
+
+	// Create the .mount unit file.
+	FILE *keyloadunit_f = fopenat(destdir_fd, mountfile,
+	    O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC, "w", 0644);
+	if (strcmp(p_canmount, "on") == 0)
+		sem_post(&noauto_files->noauto_names_sem);
+	if (!keyloadunit_f) {
+		fprintf(stderr,
+		    PROGNAME ": %s (%d): couldn't open %s under %s: %s\n",
+		    dataset, getpid(), mountfile, destdir, strerror(errno));
+		return (1 << 7);
+	}
+
+	fprintf(keyloadunit_f,
+	    OUTPUT_HEADER
+	    "[Unit]\n"
+	    "SourcePath=" FSLIST "/%s\n"
+	    "Documentation=man:zfs-mount-generator(8)\n"
+	    "\n"
+	    "Before=",
+	    cachefile);
+
+	if (p_systemd_before)
+		fprintf(keyloadunit_f, "%s ", p_systemd_before);
+	fprintf(keyloadunit_f, "zfs-mount.service");
+	if (requiredby)
+		fprintf(keyloadunit_f, " %s", requiredby);
+	if (wantedby && wantedby_append)
+		fprintf(keyloadunit_f, " %s", wantedby);
+
+	fprintf(keyloadunit_f,
+	    "\n"
+	    "After=");
+	if (p_systemd_after)
+		fprintf(keyloadunit_f, "%s ", p_systemd_after);
+	fprintf(keyloadunit_f, "%s\n", after);
+
+	fprintf(keyloadunit_f, "Wants=%s\n", wants);
+
+	if (bindsto)
+		fprintf(keyloadunit_f, "BindsTo=%s\n", bindsto);
+	if (p_systemd_requires)
+		fprintf(keyloadunit_f, "Requires=%s\n", p_systemd_requires);
+	if (p_systemd_requiresmountsfor)
+		fprintf(keyloadunit_f,
+		    "RequiresMountsFor=%s\n", p_systemd_requiresmountsfor);
+
+	fprintf(keyloadunit_f,
+	    "\n"
+	    "[Mount]\n"
+	    "Where=%s\n"
+	    "What=%s\n"
+	    "Type=zfs\n"
+	    "Options=defaults%s,zfsutil\n",
+	    p_mountpoint, dataset, opts);
+
+	(void) fclose(keyloadunit_f);
+
+	if (!requiredby && !wantedby)
+		return (0);
+
+	// Finally, create the appropriate dependencies
+	char *linktgt;
+	if (asprintf(&linktgt, "../%s", mountfile) == -1)
+		ABORT_ENOMEM();
+
+	char *dependencies[][2] = {
+		{"wants", wantedby},
+		{"requires", requiredby},
+		{}
+	};
+	for (__typeof__(&*dependencies) dep = &*dependencies; **dep; ++dep)
```

Comment:
Why `__typeof__` here?

Reply:
Do you have a better way to spell `char * (*)[2]`?

Reply:
No, I can't come up with anything cleaner at the moment.